cloudflare.go

```diff
@@ -260,30 +260,25 @@ func (api *API) makeRequestWithAuthTypeAndHeaders(ctx context.Context, method, u
 		return nil, respErr
 	}
 
-	switch {
-	case resp.StatusCode >= http.StatusOK && resp.StatusCode < http.StatusMultipleChoices:
-	case resp.StatusCode == http.StatusUnauthorized:
-		return nil, errorFromResponse(resp.StatusCode, respBody)
-	case resp.StatusCode == http.StatusForbidden:
-		return nil, errorFromResponse(resp.StatusCode, respBody)
-	case resp.StatusCode == http.StatusServiceUnavailable,
-		resp.StatusCode == http.StatusBadGateway,
-		resp.StatusCode == http.StatusGatewayTimeout,
-		resp.StatusCode == 522,
-		resp.StatusCode == 523,
-		resp.StatusCode == 524:
-		return nil, errors.Errorf("HTTP status %d: service failure", resp.StatusCode)
-	// This isn't a great solution due to the way the `default` case is
-	// a catch all and that the `filters/validate-expr` returns a HTTP 400
-	// yet the clients need to use the HTTP body as a JSON string.
-	case resp.StatusCode == 400 && strings.HasSuffix(resp.Request.URL.Path, "/filters/validate-expr"):
-		return nil, errors.Errorf("%s", respBody)
-	default:
-		var s string
-		if respBody != nil {
-			s = string(respBody)
+	if resp.StatusCode >= http.StatusBadRequest {
+		if strings.HasSuffix(resp.Request.URL.Path, "/filters/validate-expr") {
+			return nil, errors.Errorf("%s", respBody)
+		}
+
```

Comment:
I wonder if we should still raise errors on all 4xx status codes (https://api.cloudflare.com/#getting-started-responses) but with the error object populated. We could define some public Error types for all those cases mentioned in the docs and return them so the user can check against using the "new" errors methods. For example return a ErrRateLimit in case of a rate limit.
This would also make this change more backwards compatible. With the current implemention I think it would be best to release a v2 so updates will not break existing implementations.

Reply:
I'll have a think about the error types and see what I come up with. I think that will be useful for some but wrapping them all up in the same type with helper methods will be slightly more accessible than dealing with Go's types. 

As for compatibility, we haven't yet released 1.x so the API should still be considered in development with breaking changes. However, we're not using that as an excuse to break the world and instead we are going to do a minor version bump to 0.14 with release notes that we've introduced some breaking changes to keep people across it.

Reply:
I've had a think about this and I can't quite see any benefit of either approach at the moment. Let's assume Go 1.13 for simplicity, here are the two comparable cases for rate limiting.

**checking err types**

```go
if errors.Is(err, cloudflare.ClientRatedLimited) {
  // perform some backoff operation
  return
}

if err != nil {
  // .. do something else
}
```

**checking with helper methods**
 
```go
// check with helper methods
if errors.Is(err, *cloudflare.APIRequestError) && err.(*cloudflare.APIRequestError).ClientRateLimited() {
  // perform some backoff operation
}
```

The other thing to take into account here is that even if we defined an err interface for all the error types to share, we'd also need to factor in returning multiple errors at once, together which seems like it's getting into murky waters for what improvements we're trying to implement here.

At this stage, I'm inclined to leave this as is unless we can find some benefits/examples to tip the scales in favour of it.